src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,263 @@
+
+self_test_frontend::self_test_frontend(
+  model::node_id self,
+  ss::sharded<cluster::members_table>& members,
+  ss::sharded<cluster::self_test_backend>& self_be,
+  ss::sharded<rpc::connection_cache>& connections)
+  : _self(self)
+  , _state({})
+  , _previous_successful_run({})
+  , _members(members)
+  , _self_be(self_be)
+  , _connections(connections) {}
+
+ss::future<> self_test_frontend::start() { return ss::now(); }
+
+ss::future<> self_test_frontend::stop() { co_await _gate.close(); }
+
+ss::future<uuid_t>
+self_test_frontend::start_test(diskcheck_opts dto, netcheck_opts nto) {
+    auto gate_holder = _gate.hold();
+    const auto stopped_results = co_await stop_test();
+    if (!stopped_results.finished()) {
+        /// hmm what to do? maybe throw? continue with less participants
+        co_return uuid_t{}; // TODO: fix
+    }
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    start_test_request req{.id = id, .dto = dto, .nto = nto};
+    co_await invoke_on_all_nodes(
+      [req](auto handle) { return handle->start_test(req); });
+    /// TODO: Throw if something bad occurred
+    co_return id;
+}
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
+    while (!test_state.finished() && retries-- > 0) {
+        /// invoke_on_all will filter brokers that have shut down cleanly from
+        /// its participants list, retries (if occur) will only occur on
+        /// previously unreachable nodes
+        test_state = co_await invoke_on_all_nodes(
+          [](auto handle) { return handle->stop_test(); });
+        if (!test_state.finished()) {
+            vlog(
+              clusterlog.warn,
+              "Failed to sucessfully stop all outstanding jobs on all "
+              "participants: {} , trying again, retries: {}",
+              test_state,
+              retries);
+        }
+        co_await ss::sleep(1s);
+    }
+    co_return test_state;
+}
+
+ss::future<self_test_frontend::global_test_state> self_test_frontend::status() {
+    return invoke_on_all_nodes(
+      [](auto handle) { return handle->get_status(); });
+}
+
+std::vector<model::node_id>
+self_test_frontend::get_active_participants() const {
+    auto participants = _state.active_participant_ids();
+    auto brokers = participants.empty() ? _members.local().node_ids()
+                                        : participants;
+    return brokers;
+}
+
+template<typename Func>
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::invoke_on_all_nodes(Func f) {
+    auto gate_holder = _gate.hold();
+    auto brokers = get_active_participants();
+    auto local_state_vec = co_await ssx::parallel_transform(
+      brokers, [this, f = std::move(f)](const auto& node_id) {
+          ss::shared_ptr<invoke_wrapper> handle;
```

Comment:
nit: I'm curious, maybe this will be more useful in later PRs, but if this is owned and passed around by this fiber, would it make sense for it to just be a `unique_ptr`?

Reply:
nice catch, was leftover from prev impl